File: src/services/local-directory-storage-service.mjs
import { createReadStream } from "node:fs";
import {
  access,
  constants,
  lstat,
  mkdir,
  readdir,
  realpath,
  rename,
  rm,
  rmdir,
} from "node:fs/promises";
import path from "node:path";
import { StorageError } from "../domain/errors.mjs";
import {
  requireChildName,
  requirePathSegment,
  STATE_DIRECTORY,
} from "../domain/storage-route.mjs";

/**
 * Local storage implementation that keeps the Express example runnable.
 * Replace this service in production while retaining the router and security
 * middleware contract.
 */
export class LocalDirectoryStorageService {
  constructor(config, stateStore) {
    this.config = config;
    this.stateStore = stateStore;
  }

  async initialize() {
    await mkdir(this.config.storageRoot, { recursive: true, mode: 0o700 });
    this.root = await realpath(this.config.storageRoot);
  }

  async info(segments) {
    const file =
      segments.length === 0 ? this.root : await this.#existingPath(segments);
    const metadata = await lstat(file);
    if (!metadata.isFile() && !metadata.isDirectory()) {
      throw new StorageError(403, "This storage item type is not supported");
    }
    const documentPath = segments.join("/");
    const lock = await this.stateStore.currentLock(documentPath);
    return {
      path: documentPath,
      name: segments.at(-1) || this.config.rootName,
      type: metadata.isDirectory() ? "directory" : "file",
      size: metadata.isDirectory() ? 0 : metadata.size,
      readable: await permission(file, constants.R_OK),
      writable: await permission(file, constants.W_OK),
      locked: lock !== null,
      locker: lock?.owner ?? null,
      createdAt: metadata.birthtime.toISOString(),
      modifiedAt: metadata.mtime.toISOString(),
      revision: revisionFor(metadata),
    };
  }

  async list(segments) {
    const directory = await this.#directory(segments);
    const names = (await readdir(directory))
      .filter((name) => name !== STATE_DIRECTORY)
      .sort();
    if (names.length > 10_000) {
      throw new StorageError(
        413,
        "The directory contains more than 10000 entries",
      );
    }
    const entries = [];
    for (const name of names)
      entries.push(await this.info([...segments, name]));
    return { entries };
  }

  async download(segments) {
    const file = await this.#existingPath(segments);
    const metadata = await lstat(file);
    if (!metadata.isFile())
      throw new StorageError(409, "The requested item is not a file");
    return { stream: createReadStream(file), contentLength: metadata.size };
  }

  async save(segments, stagingFile) {
    if (segments.length === 0)
      throw new StorageError(400, "A document path is required");
    await this.#directory(segments.slice(0, -1));
    const target = await this.#path(segments, true);
    const existing = await lstat(target).catch((error) => {
      if (error?.code === "ENOENT") return null;
      throw error;
    });
    if (existing?.isDirectory()) {
      throw new StorageError(409, "A directory already uses this path");
    }
    await rename(stagingFile, target);
    return revisionFor(await lstat(target));
  }

  async lock(segments, owner) {
    await this.info(segments);
    await this.stateStore.acquireLock(segments.join("/"), owner);
  }

  async unlock(segments, owner) {
    await this.stateStore.releaseLock(segments.join("/"), owner);
  }

  async createDirectory(segments, name) {
    const parent = await this.#directory(segments);
    const target = path.join(parent, requireChildName(name));
    await mkdir(target, { mode: 0o700 }).catch((error) => {
      if (error?.code === "EEXIST")
        throw new StorageError(409, "An item already uses this name");
      throw error;
    });
  }

  async rename(segments, name) {
    if (segments.length === 0)
      throw new StorageError(400, "The storage root cannot be renamed");
    await this.stateStore.requireUnlocked(segments.join("/"));
    const source = await this.#existingPath(segments);
    const target = path.join(path.dirname(source), requireChildName(name));
    const exists = await lstat(target)
      .then(() => true)
      .catch((error) => {
        if (error?.code === "ENOENT") return false;
        throw error;
      });
    if (exists) throw new StorageError(409, "An item already uses this name");
    await rename(source, target);
  }

  async delete(segments) {
    if (segments.length === 0)
      throw new StorageError(400, "The storage root cannot be deleted");
    await this.stateStore.requireUnlocked(segments.join("/"));
    const target = await this.#existingPath(segments);
    const metadata = await lstat(target);
    if (metadata.isDirectory()) await rmdir(target);
    else await rm(target, { force: false });
  }

  async #directory(segments) {
    const directory =
      segments.length === 0 ? this.root : await this.#existingPath(segments);
    const metadata = await lstat(directory);
    if (!metadata.isDirectory())
      throw new StorageError(409, "The parent path is not a directory");
    return directory;
  }

  async #existingPath(segments) {
    return this.#path(segments, false);
  }

  async #path(segments, allowMissingFinal) {
    let current = this.root;
    for (let index = 0; index < segments.length; index += 1) {
      current = path.join(current, requirePathSegment(segments[index]));
      const relative = path.relative(this.root, current);
      if (relative.startsWith("..") || path.isAbsolute(relative)) {
        throw new StorageError(
          400,
          "The document path escapes the storage root",
        );
      }
      try {
        const metadata = await lstat(current);
        if (metadata.isSymbolicLink()) {
          throw new StorageError(
            403,
            "Symbolic links are not available through this Provider",
          );
        }
      } catch (error) {
        if (error instanceof StorageError) throw error;
        if (
          error?.code === "ENOENT" &&
          allowMissingFinal &&
          index === segments.length - 1
        ) {
          return current;
        }
        throw error;
      }
    }
    return current;
  }
}

async function permission(file, flag) {
  try {
    await access(file, flag);
    return true;
  } catch {
    return false;
  }
}

function revisionFor(metadata) {
  return `${Math.trunc(metadata.mtimeMs)}-${metadata.size}`;
}

File: examples/java/src/main/java/com/thinkfree/storage/service/LocalDirectoryStorageService.java
package com.thinkfree.storage.service;

import com.thinkfree.storage.config.StorageProperties;
import com.thinkfree.storage.model.StorageEntry;
import com.thinkfree.storage.model.StorageListResponse;
import org.springframework.core.io.FileSystemResource;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.DirectoryStream;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.Files;
import java.nio.file.LinkOption;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Local-filesystem storage implementation used to keep the Spring example runnable.
 * Production Providers can replace this service without changing the web or security layers.
 */
@Service
public class LocalDirectoryStorageService {
    private final StorageProperties properties;
    private final StorageStateStore stateStore;
    private final Path root;

    public LocalDirectoryStorageService(StorageProperties properties, StorageStateStore stateStore)
            throws IOException {
        this.properties = properties;
        this.stateStore = stateStore;
        Files.createDirectories(properties.root());
        root = properties.root().toRealPath();
    }

    public StorageEntry info(List<String> path) throws IOException {
        Path item = path.isEmpty() ? root : existingPath(path);
        BasicFileAttributes attributes = Files.readAttributes(
                item, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
        if (!attributes.isRegularFile() && !attributes.isDirectory()) {
            throw new StorageException(HttpStatus.FORBIDDEN, "This storage item type is not supported");
        }
        if (attributes.isRegularFile() && attributes.size() > properties.maxDocumentBytes()) {
            throw new StorageException(HttpStatus.PAYLOAD_TOO_LARGE,
                    "The document exceeds the Provider size limit");
        }
        String documentPath = String.join("/", path);
        StorageStateStore.LockStatus lock = stateStore.currentLock(documentPath);
        return new StorageEntry(
                documentPath,
                path.isEmpty() ? properties.rootName() : path.get(path.size() - 1),
                attributes.isDirectory() ? "directory" : "file",
                attributes.isDirectory() ? 0 : attributes.size(),
                Files.isReadable(item),
                Files.isWritable(item),
                lock != null,
                lock == null ? null : lock.owner(),
                attributes.creationTime().toInstant().toString(),
                attributes.lastModifiedTime().toInstant().toString(),
                revision(attributes)
        );
    }

    public StorageListResponse list(List<String> path) throws IOException {
        Path directory = directory(path);
        List<Path> children = new ArrayList<>();
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(directory,
                child -> !StorageStateStore.STATE_DIRECTORY.equals(child.getFileName().toString()))) {
            for (Path child : stream) children.add(child);
        }
        children.sort(Comparator.comparing(child -> child.getFileName().toString()));
        if (children.size() > 10_000) {
            throw new StorageException(HttpStatus.PAYLOAD_TOO_LARGE,
                    "The directory contains more than 10000 entries");
        }
        List<StorageEntry> entries = new ArrayList<>();
        for (Path child : children) {
            List<String> childPath = new ArrayList<>(path);
            childPath.add(child.getFileName().toString());
            entries.add(info(List.copyOf(childPath)));
        }
        return new StorageListResponse(List.copyOf(entries));
    }

    public Download download(List<String> path) throws IOException {
        Path file = existingPath(path);
        if (!Files.isRegularFile(file, LinkOption.NOFOLLOW_LINKS)) {
            throw new StorageException(HttpStatus.CONFLICT, "The requested item is not a file");
        }
        if (Files.size(file) > properties.maxDocumentBytes()) {
            throw new StorageException(HttpStatus.PAYLOAD_TOO_LARGE,
                    "The document exceeds the Provider size limit");
        }
        return new Download(new FileSystemResource(file), Files.size(file));
    }

    public String save(List<String> path, Path stagedFile) throws IOException {
        if (path.isEmpty()) {
            throw new StorageException(HttpStatus.BAD_REQUEST, "A document path is required");
        }
        directory(path.subList(0, path.size() - 1));
        Path target = path(path, true);
        if (Files.isDirectory(target, LinkOption.NOFOLLOW_LINKS)) {
            throw new StorageException(HttpStatus.CONFLICT, "A directory already uses this path");
        }
        try {
            Files.move(stagedFile, target, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
        } catch (AtomicMoveNotSupportedException exception) {
            Files.move(stagedFile, target, StandardCopyOption.REPLACE_EXISTING);
        }
        return revision(Files.readAttributes(target, BasicFileAttributes.class));
    }

    public void lock(List<String> path, String owner) throws IOException {
        info(path);
        stateStore.lock(String.join("/", path), owner);
    }

    public void unlock(List<String> path, String owner) throws IOException {
        stateStore.unlock(String.join("/", path), owner);
    }

    public void createDirectory(List<String> path, String name) throws IOException {
        Path target = directory(path).resolve(StoragePathPolicy.requireChildName(name));
        try {
            Files.createDirectory(target);
        } catch (FileAlreadyExistsException exception) {
            throw new StorageException(HttpStatus.CONFLICT, "An item already uses this name");
        }
    }

    public void rename(List<String> path, String name) throws IOException {
        if (path.isEmpty()) {
            throw new StorageException(HttpStatus.BAD_REQUEST, "The storage root cannot be renamed");
        }
        String documentPath = String.join("/", path);
        stateStore.requireUnlocked(documentPath);
        Path source = existingPath(path);
        Path target = source.getParent().resolve(StoragePathPolicy.requireChildName(name));
        if (Files.exists(target, LinkOption.NOFOLLOW_LINKS)) {
            throw new StorageException(HttpStatus.CONFLICT, "An item already uses this name");
        }
        Files.move(source, target);
    }

    public void delete(List<String> path) throws IOException {
        if (path.isEmpty()) {
            throw new StorageException(HttpStatus.BAD_REQUEST, "The storage root cannot be deleted");
        }
        stateStore.requireUnlocked(String.join("/", path));
        Files.delete(existingPath(path));
    }

    private Path directory(List<String> path) throws IOException {
        Path directory = path.isEmpty() ? root : existingPath(path);
        if (!Files.isDirectory(directory, LinkOption.NOFOLLOW_LINKS)) {
            throw new StorageException(HttpStatus.CONFLICT, "The parent path is not a directory");
        }
        return directory;
    }

    private Path existingPath(List<String> path) throws IOException {
        return path(path, false);
    }

    private Path path(List<String> segments, boolean allowMissingFinal) throws IOException {
        Path current = root;
        for (int index = 0; index < segments.size(); index++) {
            current = current.resolve(StoragePathPolicy.requireSegment(segments.get(index))).normalize();
            if (!current.startsWith(root)) {
                throw new StorageException(HttpStatus.BAD_REQUEST, "The path escapes the storage root");
            }
            if (Files.exists(current, LinkOption.NOFOLLOW_LINKS)) {
                if (Files.isSymbolicLink(current)) {
                    throw new StorageException(HttpStatus.FORBIDDEN,
                            "Symbolic links are not available through this Provider");
                }
            } else if (!(allowMissingFinal && index == segments.size() - 1)) {
                throw new NoSuchFileException(current.toString());
            }
        }
        return current;
    }

    private static String revision(BasicFileAttributes attributes) {
        return attributes.lastModifiedTime().toMillis() + "-" + attributes.size();
    }

    public record Download(FileSystemResource resource, long contentLength) {}
}
